File: backend/app/schemas/layout.py
from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.discover import Pagination


class AssetType(str, Enum):
    COVER_IMAGE = "cover_image"
    CONTENT_IMAGE = "content_image"


class LayoutFontFamily(str, Enum):
    SANS = "sans"
    SERIF = "serif"
    ROUND = "round"


class LayoutTitleAlign(str, Enum):
    LEFT = "left"
    CENTER = "center"


class LayoutDraftStatus(str, Enum):
    DRAFT = "draft"
    PUBLISHED = "published"


class LayoutRenderRequest(BaseModel):
    title: str = Field(default="", max_length=255)
    article_id: str | None = None
    cover_asset_id: str | None = None
    content_md: str = Field(min_length=1)
    theme_id: str = Field(default="default", max_length=32)
    theme_color: str = Field(default="#FF6600", max_length=16)
    font_family: LayoutFontFamily = LayoutFontFamily.SANS
    font_size: int = Field(default=15, ge=12, le=20)
    line_height: float = Field(default=1.8, ge=1.0, le=3.0)
    title_align: LayoutTitleAlign = LayoutTitleAlign.CENTER
    para_indent: bool = True
    round_image: bool = True


class LayoutRenderResponse(BaseModel):
    html: str


class LayoutDraftUpsertRequest(LayoutRenderRequest):
    pass


class LayoutDraftItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    article_id: str | None = None
    title: str
    cover_asset_id: str | None = None
    content_md: str
    content_html: str
    theme_id: str
    theme_color: str
    font_family: LayoutFontFamily
    font_size: int
    line_height: float
    title_align: LayoutTitleAlign
    para_indent: bool
    round_image: bool
    status: LayoutDraftStatus
    created_at: datetime
    updated_at: datetime


class LayoutDraftList(BaseModel):
    items: list[LayoutDraftItem]
    pagination: Pagination


class AssetItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    asset_type: AssetType
    file_name: str
    mime_type: str
    size: int
    public_url: str

File: frontend/src/lib/layout.ts
import { requestAuthedJson } from "./auth";
import type { Pagination } from "./discover";

export type LayoutFontFamily = "sans" | "serif" | "round";
export type LayoutTitleAlign = "left" | "center";
export type LayoutDraftStatus = "draft" | "published";
export type AssetType = "cover_image" | "content_image";

export type LayoutPayload = {
  title: string;
  article_id?: string | null;
  cover_asset_id?: string | null;
  content_md: string;
  theme_id: string;
  theme_color: string;
  font_family: LayoutFontFamily;
  font_size: number;
  line_height: number;
  title_align: LayoutTitleAlign;
  para_indent: boolean;
  round_image: boolean;
};

export type LayoutRenderResponse = {
  html: string;
};

export type LayoutDraft = {
  id: string;
  article_id: string | null;
  title: string;
  cover_asset_id: string | null;
  content_md: string;
  content_html: string;
  theme_id: string;
  theme_color: string;
  font_family: LayoutFontFamily;
  font_size: number;
  line_height: number;
  title_align: LayoutTitleAlign;
  para_indent: boolean;
  round_image: boolean;
  status: LayoutDraftStatus;
  created_at: string;
  updated_at: string;
};

export type LayoutDraftList = {
  items: LayoutDraft[];
  pagination: Pagination;
};

export type Asset = {
  id: string;
  asset_type: AssetType;
  file_name: string;
  mime_type: string;
  size: number;
  public_url: string;
};

export async function renderLayout(payload: LayoutPayload) {
  return requestAuthedJson<LayoutRenderResponse>("/api/v1/layout/render", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function createLayoutDraft(payload: LayoutPayload) {
  return requestAuthedJson<LayoutDraft>("/api/v1/layout/drafts", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateLayoutDraft(draftId: string, payload: LayoutPayload) {
  return requestAuthedJson<LayoutDraft>(`/api/v1/layout/drafts/${draftId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function fetchLayoutDraft(draftId: string) {
  return requestAuthedJson<LayoutDraft>(`/api/v1/layout/drafts/${draftId}`);
}

export async function fetchLayoutDrafts(params: { page?: number; page_size?: number } = {}) {
  const searchParams = new URLSearchParams();
  if (params.page !== undefined) {
    searchParams.set("page", String(params.page));
  }
  if (params.page_size !== undefined) {
    searchParams.set("page_size", String(params.page_size));
  }
  const queryString = searchParams.toString();
  return requestAuthedJson<LayoutDraftList>(
    `/api/v1/layout/drafts${queryString ? `?${queryString}` : ""}`,
  );
}

export async function uploadAsset(file: File, assetType: AssetType) {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("asset_type", assetType);

  return requestAuthedJson<Asset>("/api/v1/assets", {
    method: "POST",
    body: formData,
  });
}
